### PR TITLE
Implement C++23 Algorithms: fold_left, fold_right.

### LINKED ISSUE #6908: Implement missing C++23 range algorithms: fold, chunk_by, slide, stride, cartesian_product
Proposal: Implement Missing C++23 Range Algorithms and Views
===============================================================================

Overview
-------------------------------------------------------------------------------

I would like to work on implementing several C++23 range algorithms that are
currently missing in HPX.

This will help move HPX closer to full C++23 standard compliance and provide
users with additional modern tools for working with ranges.

Planned Additions
===============================================================================

1) Fold Algorithms (P2322R6)
-------------------------------------------------------------------------------

ranges::fold_left family:
    - fold_left
    - fold_left_with_iter
    - fold_left_first
    - fold_left_first_with_iter

ranges::fold_right family:
    - fold_right
    - fold_right_last

2) Range Adaptors / Views
-------------------------------------------------------------------------------

- views::chunk_by        (P2443R1)
- views::slide           (P2442R1)
- views::stride          (P1899R3)
- views::cartesian_product (P2374R4)
- views::repeat          (P2474R2)

Implementation Plan
===============================================================================

- Implement fold algorithms as Customization Point Objects (CPOs)
  using hpx::tag_invoke, following the pattern used in reduce.

- Place implementations under:
    hpx/parallel/container_algorithms   (for fold algorithms)
    hpx/parallel/util                   (for views)

- Ensure support for:
    - sequenced execution policy
    - parallel execution policy
    - parallel unsequenced execution policy

- Add unit tests for each algorithm/view covering:
    - Basic correctness
    - Sequential execution
    - Parallel execution
    - Edge cases (e.g., empty ranges)

### PR DESCRIPTION
**Pull request by arpittkhandelwal**
fix #6908

## Description
This PR implements several missing C++23 algorithms and range adaptors in `hpx::parallel`.

### Added Algorithms
- [fold_left](cci:1://file:///Users/arpitkhandelwal/.gemini/antigravity/scratch/hpx/libs/core/algorithms/include/hpx/parallel/container_algorithms/fold_left.hpp:16:4-18:49) family: [fold_left](cci:1://file:///Users/arpitkhandelwal/.gemini/antigravity/scratch/hpx/libs/core/algorithms/include/hpx/parallel/container_algorithms/fold_left.hpp:16:4-18:49), [fold_left_with_iter](cci:1://file:///Users/arpitkhandelwal/.gemini/antigravity/scratch/hpx/libs/core/algorithms/include/hpx/parallel/container_algorithms/fold_left.hpp:20:4-22:59), [fold_left_first](cci:1://file:///Users/arpitkhandelwal/.gemini/antigravity/scratch/hpx/libs/core/algorithms/include/hpx/parallel/container_algorithms/fold_left.hpp:24:4-26:47), [fold_left_first_with_iter](cci:1://file:///Users/arpitkhandelwal/.gemini/antigravity/scratch/hpx/libs/core/algorithms/tests/unit/container_algorithms/fold_left_range.cpp:59:0-66:1)
- [fold_right](cci:1://file:///Users/arpitkhandelwal/.gemini/antigravity/scratch/hpx/libs/core/algorithms/include/hpx/parallel/container_algorithms/fold_right.hpp:16:4-18:50) family: [fold_right](cci:1://file:///Users/arpitkhandelwal/.gemini/antigravity/scratch/hpx/libs/core/algorithms/include/hpx/parallel/container_algorithms/fold_right.hpp:16:4-18:50), [fold_right_last](cci:1://file:///Users/arpitkhandelwal/.gemini/antigravity/scratch/hpx/libs/core/algorithms/include/hpx/parallel/container_algorithms/fold_right.hpp:20:4-22:47)

### Implementation Details
- **Sequential Implementation**: As per discussion, these are currently implemented sequentially.
- **Export Macros**: All CPOs and helper structs use `HPX_CXX_CORE_EXPORT` for proper visibility.
- **Utils**: Added `in_value_result` to [hpx/parallel/util/result_types.hpp](cci:7://file:///Users/arpitkhandelwal/.gemini/antigravity/scratch/hpx/libs/core/algorithms/include/hpx/parallel/util/result_types.hpp:0:0-0:0).

## Verification
- Added unit tests in `libs/core/algorithms/tests/unit/container_algorithms/`.
- All tests passed locally.

### CODE REVIEW

**Comment by hkaiser** [comment]
@arpittkhandelwal before I start reviewing all of your contributions (thanks for working on this) - one question: what would be the benefit of re-implementing any of the range views in HPX if the standard already provides those (apart from the fact that those are C++23 only). Do we need those views for any of our own code at this time?

For the algorithms (even the sequential only versions) I can still see some minimal benefit as people may not want to move to C++23 just for being able to use one of the `fold`s. But if somebody really wants to use ranges to an extent that they need advanced views like the `cartesian_product`, I think they would be using C++23 already, thus wouldn't need us to provide such an implementation.

May I suggest to separate the fold algorithms from the views for now? I'm a bit hesitant to add a lot of functionality that should be available elsewhere.

**Comment by arpittkhandelwal** [comment]
> @arpittkhandelwal before I start reviewing all of your contributions (thanks for working on this) - one question: what would be the benefit of re-implementing any of the range views in HPX if the standard already provides those (apart from the fact that those are C++23 only). Do we need those views for any of our own code at this time?
> 
> For the algorithms (even the sequential only versions) I can still see some minimal benefit as people may not want to move to C++23 just for being able to use one of the `fold`s. But if somebody really wants to use ranges to an extent that they need advanced views like the `cartesian_product`, I think they would be using C++23 already, thus wouldn't need us to provide such an implementation.
> 
> May I suggest to separate the fold algorithms from the views for now? I'm a bit hesitant to add a lot of functionality that should be available elsewhere.


Thank you for the feedback sir,that’s a very fair point.

I agree that for the views, the benefit is less clear since they are already available in C++23. My initial thinking was mainly around consistency and offering a unified HPX interface, especially for users who may still be on pre-C++23 toolchains. But I understand the concern about duplicating functionality that the standard library already provides.

For the fold algorithms, I agree there’s clearer value especially since they integrate with execution policies and provide C++23 functionality without requiring users to move to C++23.

I’m happy to split this PR and focus only on the fold algorithms for now, and we can revisit the views later if there’s a stronger use case or internal need.

File: libs/core/algorithms/include/hpx/parallel/container_algorithms/fold_left.hpp [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+/// \page hpx::ranges::fold_left
+/// \headerfile hpx/algorithm.hpp
+
+#pragma once
+
+#if defined(DOXYGEN)
+namespace hpx { namespace ranges {
+    // clang-format off
+
+    /// Loose implementation of the C++23 fold_left algorithm
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/parallel/container_algorithms/fold_left.hpp]: What do you mean by `loose`?

File: libs/core/algorithms/include/hpx/parallel/container_algorithms/fold_left.hpp [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+namespace hpx::ranges::detail {
+    template <typename I, typename S, typename T, typename F>
+    constexpr auto fold_left_with_iter_impl(I first, S last, T init, F f)
+    {
+        using U = std::decay_t<T>;
+        if (first == last)
+        {
+            return hpx::parallel::util::in_value_result<I, U>{
+                std::move(first), std::move(init)};
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/parallel/container_algorithms/fold_left.hpp]: Inside HPX, please use `HPX_MOVE` instead of `std::move`. This reduces compile times:
```suggestion
                HPX_MOVE(first), HPX_MOVE(init)};
```

File: libs/core/algorithms/include/hpx/parallel/container_algorithms/fold_left.hpp [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+    }
+}    // namespace hpx::ranges::detail
+
+namespace hpx::ranges {
+
+    // ----------------------------------------------------------------------
+    // fold_left_with_iter
+    // ----------------------------------------------------------------------
+    HPX_CXX_CORE_EXPORT struct fold_left_with_iter_t final
+      : hpx::functional::detail::tag_fallback<fold_left_with_iter_t>
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/parallel/container_algorithms/fold_left.hpp]: Should we derive the CPOs from 
```suggestion
      : hpx::detail::tag_parallel_algorithm<fold_left_with_iter_t>
```
This will integrate them with senders&receivers, amongst other things.

File: libs/core/algorithms/include/hpx/parallel/container_algorithms/fold_right.hpp [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+#include <type_traits>
+#include <utility>
+
+namespace hpx::ranges::detail {
+    template <typename I, typename S, typename T, typename F>
+    constexpr auto fold_right_impl(I first, S last, T init, F f)
+    {
+        using U = std::decay_t<T>;
+
+        auto end_iter = hpx::ranges::next(first, last);
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/parallel/container_algorithms/fold_right.hpp]: Will this do the right thing even if `S` is a sentinel (not an iterator)? Would it be better to write:
```suggestion
        auto end_iter = detail::advance_to_sentinel(first, last);
```

File: libs/core/algorithms/include/hpx/parallel/util/result_types.hpp [hunk 5]
```diff
@@ -400,10 +398,43 @@ namespace hpx::parallel::util {
...
+    struct in_value_result
+    {
+        HPX_NO_UNIQUE_ADDRESS I in;
+        HPX_NO_UNIQUE_ADDRESS T value;
+
+        template <typename I2, typename T2,
+            typename Enable =
+                std::enable_if_t<std::is_convertible_v<I const&, I2> &&
+                    std::is_convertible_v<T const&, T2>>>
+        constexpr operator in_value_result<I2, T2>() const&
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/parallel/util/result_types.hpp]: Let's use concepts, if possible:
```suggestion
        template <typename I2, typename T2>
            requires(std::convertible_to<I const&, I2> && std::convertible_to<T const&, T2>)
        constexpr operator in_value_result<I2, T2>() const&
```

File: libs/core/algorithms/tests/unit/container_algorithms/fold_left_range.cpp [hunk 6]
```diff
@@ -0,0 +1,79 @@
+//  Copyright (c) 2026 Arpit Khandelwal
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#include <hpx/algorithm.hpp>
+#include <hpx/init.hpp>
+#include <hpx/modules/testing.hpp>
+#include <hpx/parallel/container_algorithms/fold_left.hpp>
```
Comment (hkaiser) [inline libs/core/algorithms/tests/unit/container_algorithms/fold_left_range.cpp]: This is not needed here. It will be included by the `<algorithm>` header

File: libs/core/algorithms/include/hpx/parallel/container_algorithms/fold_left.hpp (hunk 1 above)
Comment (arpittkhandelwal) [reply libs/core/algorithms/include/hpx/parallel/container_algorithms/fold_left.hpp]: Thanks sir for the catch! The 'Loose' description was a holdover from my initial draft where I wasn't fully using concepts. I've removed it now as the implementation is intended to be a robust C++23 implementation within HPX constraints

File: libs/core/algorithms/include/hpx/parallel/container_algorithms/fold_left.hpp (hunk 2 above)
Comment (arpittkhandelwal) [reply libs/core/algorithms/include/hpx/parallel/container_algorithms/fold_left.hpp]: I've updated the code to use HPX_MOVE consistently throughout the implementation.

File: libs/core/algorithms/include/hpx/parallel/container_algorithms/fold_left.hpp (hunk 3 above)
Comment (arpittkhandelwal) [reply libs/core/algorithms/include/hpx/parallel/container_algorithms/fold_left.hpp]: Done. I've updated the CPOs to derive from hpx::parallel::util::detail::tag_parallel_algorithm. This should ensure better integration with senders/receivers.

File: libs/core/algorithms/include/hpx/parallel/container_algorithms/fold_right.hpp (hunk 4 above)
Comment (arpittkhandelwal) [reply libs/core/algorithms/include/hpx/parallel/container_algorithms/fold_right.hpp]: Yes, you are right. hpx::ranges::next might not handle sentinels correctly in all cases. I've already updated the implementation to use hpx::parallel::detail::advance_to_sentinel(first, last) which is designed exactly for this purpose.

File: libs/core/algorithms/include/hpx/parallel/util/result_types.hpp (hunk 5 above)
Comment (arpittkhandelwal) [reply libs/core/algorithms/include/hpx/parallel/util/result_types.hpp]: Done. I've updated in_value_result and all other result types in this file to use C++20 requires clauses with std::convertible_to

File: libs/core/algorithms/tests/unit/container_algorithms/fold_left_range.cpp (hunk 6 above)
Comment (arpittkhandelwal) [reply libs/core/algorithms/tests/unit/container_algorithms/fold_left_range.cpp]: Done. I've removed the explicit include.

**Comment by hkaiser** [comment]
Some of your changes overlap with https://github.com/STEllAR-GROUP/hpx/pull/6884, please coordinate with @sleepingeight. This PR also introduces some of the necessary concepts which you could then use to add concept checking to the new algorithms you introduced.

**Comment by arpittkhandelwal** [comment]
> Some of your changes overlap with #6884, please coordinate with @sleepingeight. This PR also introduces some of the necessary concepts which you could then use to add concept checking to the new algorithms you introduced

Thank you sir for pointing this out.
I see the overlap with #6884. I’ll coordinate with @sleepingeight, rebase on their changes, and reuse the introduced concepts instead of duplicating anything.

I’ll resolve the conflicts and update the PR once aligned.

**Comment by arpittkhandelwal** [comment]
@sleepingeight  Hi! I see that your PR (#6884) introduces fold concepts and traits, and my PR (#6910) also implements fold algorithms.  

Since there’s some overlap, how would you like to coordinate? Should I rebase on your branch and reuse your concepts?

**Comment by sleepingeight** [comment]
HI @arpittkhandelwal, this PR seems to be very similar to my one. I guess you could use the concepts I've implemented in `is_foldable.hpp` file. Also the parameters of the algorithms need to satisfy many requirements. You could take a look at my PR and implement as it is.

**Comment by arpittkhandelwal** [comment]
> HI @arpittkhandelwal, this PR seems to be very similar to my one. I guess you could use the concepts I've implemented in `is_foldable.hpp` file. Also the parameters of the algorithms need to satisfy many requirements. You could take a look at my PR and implement as it is.

Thanks @sleepingeight for the suggestion! I've updated the implementation to use the fold concepts from your PR (is_foldable.hpp) in fold_left and fold_right algorithms. I also squashed these changes into the main commit.

**Comment by arpittkhandelwal** [comment]
> > Some of your changes overlap with #6884, please coordinate with @sleepingeight. This PR also introduces some of the necessary concepts which you could then use to add concept checking to the new algorithms you introduced
> 
> Thank you sir for pointing this out. I see the overlap with #6884. I’ll coordinate with @sleepingeight, rebase on their changes, and reuse the introduced concepts instead of duplicating anything.
> 
> I’ll resolve the conflicts and update the PR once aligned.

Sir I updated my pr according to the 6884 pr

**Comment by hkaiser** [comment]
> Sir I updated my pr according to the 6884 pr

Let's see for #6884 to be merged first, then you can update your code with whatever latest version of the concepts @sleepingeight will have come up with.

**Comment by arpittkhandelwal** [comment]
> > Sir I updated my pr according to the 6884 pr
> 
> Let's see for #6884 to be merged first, then you can update your code with whatever latest version of the concepts @sleepingeight will have come up with.

As PR #6884 has now been successfully merged into master, I have rebased this branch and updated the fold algorithms implementation to use the latest version of the concepts introduced by @sleepingeight. 

The implementation now perfectly aligns with the upstream structure. Please let me know if there are any further changes required!

File: libs/core/algorithms/include/hpx/parallel/container_algorithms/fold.hpp [hunk 7]
```diff
@@ -0,0 +1,580 @@
...
+        {
+            using U = std::decay_t<hpx::util::invoke_result_t<F&,
+                hpx::traits::iter_reference_t<BidIter>, T>>;
+
+            if (first == last)
+            {
+                return U(HPX_MOVE(init));
+            }
+
+            auto it = hpx::ranges::next(first, last);
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/parallel/container_algorithms/fold.hpp]: I think this should be using `advance_to_sentinel` instead of `ranges::next`. The same is true for other uses of `ranges::next` below.

File: libs/core/iterator_support/include/hpx/iterator_support/traits/is_foldable.hpp [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#pragma once
+
+#include <hpx/config.hpp>
+#include <hpx/functional/invoke_result.hpp>
+#include <hpx/functional/traits/is_invocable.hpp>
```
Comment (hkaiser) [inline libs/core/iterator_support/include/hpx/iterator_support/traits/is_foldable.hpp]:
```suggestion
#include <hpx/modules/functional.hpp>
```

File: libs/core/iterator_support/include/hpx/iterator_support/traits/is_foldable.hpp [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+#include <hpx/iterator_support/boost_iterator_categories.hpp>
+#include <hpx/modules/type_support.hpp>
+#include <concepts>
```
Comment (hkaiser) [inline libs/core/iterator_support/include/hpx/iterator_support/traits/is_foldable.hpp:15]:
```suggestion

#include <concepts>
```

**Comment by arpittkhandelwal** [comment]
@hkaiser sir Done! I've addressed all three review comments:
1. Replaced  `hpx::ranges::next` with `hpx::parallel::detail::advance_to_sentinel` in both `fold_right` and `fold_right_last`  
2. Consolidated `hpx/functional/invoke_result.hpp` + `hpx/functional/traits/is_invocable.hpp` into `hpx/modules/functional.hpp`  
3. Moved `<concepts>` to after the HPX includes

**Comment by hkaiser** [comment]
> @hkaiser sir Done! I've addressed all three review comments:
> 
> 1. Replaced  `hpx::ranges::next` with `hpx::parallel::detail::advance_to_sentinel` in both `fold_right` and `fold_right_last`
> 2. Consolidated `hpx/functional/invoke_result.hpp` + `hpx/functional/traits/is_invocable.hpp` into `hpx/modules/functional.hpp`
> 3. Moved `<concepts>` to after the HPX includes

Thanks! Could you please rebase to resolve the conflicts? After that we can merge.

**Comment by arpittkhandelwal** [comment]
> > @hkaiser sir Done! I've addressed all three review comments:
> > 
> > 1. Replaced  `hpx::ranges::next` with `hpx::parallel::detail::advance_to_sentinel` in both `fold_right` and `fold_right_last`
> > 2. Consolidated `hpx/functional/invoke_result.hpp` + `hpx/functional/traits/is_invocable.hpp` into `hpx/modules/functional.hpp`
> > 3. Moved `<concepts>` to after the HPX includes
> 
> Thanks! Could you please rebase to resolve the conflicts? After that we can merge.

@hkaiser sir  I've successfully merged the latest changes from `upstream/master` and resolved the conflicts in [fold.hpp](cci:7://file:///Users/arpitkhandelwal/.gemini/antigravity/scratch/hpx/libs/core/algorithms/include/hpx/parallel/container_algorithms/fold.hpp:0:0-0:0) and [is_foldable.hpp](cci:7://file:///Users/arpitkhandelwal/.gemini/antigravity/scratch/hpx/libs/core/iterator_support/include/hpx/iterator_support/traits/is_foldable.hpp:0:0-0:0) (incorporating both the review fixes and the recent upstream concept updates).

The tests run successfully, and the PR should now be clear to merge!

**Comment by arpittkhandelwal** [comment]
@hkaiser Sir anything left in this?
